src/handlers.rs

```diff
@@ -393,6 +397,147 @@ where
+                .map_err(|e| Error::Generic(format!("Unable to build transaction: {}", e)))?;
+
+            // check to make sure it is finalized (sign will return bool true )
+
+            // set signoptions
+            let mut sign_options = SignOptions::default();
+            sign_options.trust_witness_utxo = true;
+
+            // check + sign psbt
+            let finalized = wallet.sign(&mut psbt, sign_options)?;
+
+            // need to save the utxo (input maps txout) here. maybe can use hashmap
+            let original_inputs = psbt.inputs.clone();
+
+            // handle error if the psbt is not finalized
+            if !finalized {
+                return Err(Error::Generic("PSBT not finalized".to_string()));
+            }
+
+            // set payjoin params
+            let pj_params = payjoin::send::Configuration::with_fee_contribution(
+                payjoin::bitcoin::Amount::from_sat(100),
+                None,
+            );
+
+            // Construct the request with the PSBT and parameters
+            let (req, ctx) = uri
+                .create_pj_request(psbt, pj_params) // not clone here
+                .map_err(|e| Error::Generic(format!("Error building payjoin request: {}", e)))?;
+
+            // reqwest client
+            let client = reqwest::blocking::Client::builder()
+                .danger_accept_invalid_certs(true)
+                .build()
+                .map_err(|e| Error::Generic(format!("Error building payjoin client: {}", e)))?;
+
+            // send the request, and receive response
+            let response = client
+                .post(req.url)
+                .body(req.body)
+                .header("Content-Type", "text/plain")
+                .send()
+                .map_err(|e| Error::Generic(format!("Error with HTTP request: {}", e)))?;
+
+            // process the response
+            let mut psbt = ctx
+                .process_response(response)
+                .map_err(|e| Error::Generic(format!("Error processing payjoin response: {}", e)))?;
+
+            // set this to account for original_inputs being of different length than psbt.inputs
+            let mut og = 0;
+
+            // Iterate over the inputs in the PSBT
+            for input in psbt.inputs.iter_mut() {
+                // If this input is some skip it
+                if input.witness_utxo.is_some() || input.non_witness_utxo.is_some() {
+                    continue;
+                }
+                // If index is within the range of original_inputs
+                else if og < original_inputs.len() {
+                    // Reintroduce the utxo from the original_inputs
+                    input.witness_utxo = original_inputs[og].witness_utxo.clone();
+                    input.non_witness_utxo = original_inputs[og].non_witness_utxo.clone();
+
+                    // increment original_index only when an original input is used
+                    og += 1;
+                }
+            }
```

Comment:
We should be able to iterate without explicit indexing as in [`process_response`'s `check_inputs`](https://github.com/payjoin/rust-payjoin/blob/21a3fd86f30909a349f447f7d112e62335e646e1/payjoin/src/send/mod.rs#L386-L463)

I suggest renaming `psbt` to `original_psbt` and `payjoin_psbt` respectively. This snippet also requires you to pass a cloned psbt to `create_pj_request` to keep the original around. It clones slightly more bytes for imo a more readable, maintainable piece of code. I've opted to clone the maps instead of dealing with mutable references in `input_pairs` for the same reason. I justify this tradeoff because I imagine this BDK-CLI command will be used for individual wallets and sends, and not high throughput enterprise send-a-million-payjoins-at-once scenarios. PSBTv2 makes this pairing nonsense go away and can't come soon enough!

Payjoin's internal `PsbtExt`'s helpers like this duplicated/modified `input_pairs` function may be worth exposing publicly in PDK if not `rust-bitcoin` or `BDK`.

```suggestion
            fn input_pairs(psbt: &PartiallySignedTransaction) -> Box<dyn Iterator<Item = (bdk::bitcoin::TxIn, Input)> + '_> {
                Box::new(
                    psbt.unsigned_tx
                        .input
                        .iter()
                        .cloned()   // Clone each TxIn for better ergonomics than &muts
                        .zip(psbt.inputs.iter().cloned())  // Clone each Input too
                )
            }
            
            let mut original_inputs = input_pairs(&original_psbt).peekable();
            
            for (proposed_txin, mut proposed_psbtin) in input_pairs(&payjoin_psbt) {
                if let Some((original_txin, original_psbtin)) = original_inputs.peek() {
                    if proposed_txin.previous_output == original_txin.previous_output {
                        proposed_psbtin.witness_utxo = original_psbtin.witness_utxo.clone();
                        proposed_psbtin.non_witness_utxo = original_psbtin.non_witness_utxo.clone();
                    }
                    original_inputs.next();
                }
            }
```

Reply:
Yes, for a long time I wanted to move parts from payjoin to rust-bitcoin. This looks like one of them but I don't remember for sure.